src/common/SurgeStorage.cpp

```diff
@@ -279,6 +280,36 @@ void SurgeStorage::refresh_patchlist()
    refreshPatchlistAddDir(false, "patches_3rdparty");
    firstUserCategory = patch_category.size();
    refreshPatchlistAddDir(true, "");
+
+   patchOrdering = std::vector<int>(patch_list.size());
+   std::iota(patchOrdering.begin(), patchOrdering.end(), 0);
+
+   auto patchCompare =
+      [this](const int &i1, const int &i2) -> bool
+      {
+         return _stricmp(patch_list[i1].name.c_str(),
+                         patch_list[i2].name.c_str()) < 0;
+      };
+
+   std::sort(patchOrdering.begin(), patchOrdering.end(), patchCompare);
+
+   patchCategoryOrdering = std::vector<int>(patch_category.size());
+   std::iota(patchCategoryOrdering.begin(), patchCategoryOrdering.end(), 0);
+
+   auto categoryCompare =
+      [this](const int &i1, const int &i2) -> bool
+      {
+         return _stricmp(patch_category[i1].name.c_str(),
+                         patch_category[i2].name.c_str()) < 0;
+      };
+
+   int groups[4] = {0, firstThirdPartyCategory, firstUserCategory,
+                    (int)patch_category.size()};
+
+   for (int i = 0; i < 3; i++)
+      std::sort(std::next(patchCategoryOrdering.begin(), groups[i]),
+                std::next(patchCategoryOrdering.begin(), groups[i + 1]),
+                categoryCompare);
```

Comment:
I supposed you could also `patchCategoryOrdering.begin() + groups[i]` but for me that just looks unholy and evil.